### PR TITLE
Pitest blog post

### PR DESCRIPTION
**Pull request by Tomabana**

### CODE REVIEW

File: _posts/2016-05-30-pitest-with-java.md [hunk 1]
```diff
@@ -0,0 +1,263 @@
+---
+layout: post
+title:  "Test your tests with pitest"
+date:   2016-05-30 08:00:00 +0200
+author: "Tomasz Banaś"
+tags: java tests
+---
+
+There is no doubt that unit tests are necessary to keep good quality of code. People used to check line coverage to determine if they have good tests, but have you ever wonder how to test your tests and check if they really tests something?
```
Comment (natif) [inline _posts/2016-05-30-pitest-with-java.md]: 1. but... Have you ever wondered.
2. really test

File: _posts/2016-05-30-pitest-with-java.md [hunk 2]
```diff
@@ -0,0 +1,263 @@
...
+In this post I'll show show you why line coverage is a bad metric and how to use mutation testing to test unit tests.
```
Comment (natif) [inline _posts/2016-05-30-pitest-with-java.md]: 1. I will show (twice used word show)
2. mutation testing. - without "to test unit tests".

File: _posts/2016-05-30-pitest-with-java.md [hunk 3]
```diff
@@ -0,0 +1,263 @@
...
+
+* changing return values of method calls:
+
+|Return type |Mutation|
+|:-------------:|:-----:|
+|boolean|change true to false, false to true|
+|int byte short|change 0 to 1, otherwise return 0|
+|long|change x to x+1|
+
+There are few libraries which can be used to mutation testing. I chose [pitest](http://pitest.org/) library.
```
Comment (natif) [inline _posts/2016-05-30-pitest-with-java.md:50]: Maybe we should mention that for java purposes you used pitest?

File: _posts/2016-05-30-pitest-with-java.md [hunk 4]
```diff
@@ -0,0 +1,263 @@
...
+
+# Usage
+I wrote small program to calculate ticket price for passengers, which will be used to demonstrate how to use mutation testing to create good tests.
```
Comment (natif) [inline _posts/2016-05-30-pitest-with-java.md]: 1. I wrote a
2. in order to create

File: _posts/2016-05-30-pitest-with-java.md [hunk 5]
```diff
@@ -0,0 +1,263 @@
...
+Program has two classes. *Passenger* class which represents traveler and *TicketPriceCalculator* which contains logic to calculate ticket price.
```
Comment (natif) [inline _posts/2016-05-30-pitest-with-java.md]: has two classes, (coma instead of dot).

File: _posts/2016-05-30-pitest-with-java.md [hunk 6]
```diff
@@ -0,0 +1,263 @@
...
+            result = totalPrice;
+        }
+
+        return result;
+    }
+}
+{% endhighlight %}
+
+
+There are three basic scenarios to cover in tests: ticket for adult, for child and for family.
```
Comment (natif) [inline _posts/2016-05-30-pitest-with-java.md]: ticket for adult, child and family.

File: _posts/2016-05-30-pitest-with-java.md [hunk 7]
```diff
@@ -0,0 +1,263 @@
...
+        passengers.add(childPassenger3);
+        passengers.add(childPassenger4);
+        double price = calculator.calculatePrice(passengers, ADULT_TICKET_PRICE, CHILD_TICKER_PRICE);
+        assertEquals((2 * ADULT_TICKET_PRICE + 2 * CHILD_TICKER_PRICE) *
+                (1 - TicketPriceCalculator.FAMILY_DISCOUNT), price, 0);
+    }
+}
+{% endhighlight %}
+
+Ok, let's run pitest to see how good are our tests.
```
Comment (natif) [inline _posts/2016-05-30-pitest-with-java.md]: how good our tests are.

File: _posts/2016-05-30-pitest-with-java.md [hunk 8]
```diff
@@ -0,0 +1,263 @@
...
+
+To run pitest we can use maven command:
```
Comment (natif) [inline _posts/2016-05-30-pitest-with-java.md]: - use is enough, without we can

File: _posts/2016-05-30-pitest-with-java.md [hunk 9]
```diff
@@ -0,0 +1,263 @@
...
+{% highlight plain %}
+mvn org.pitest:pitest-maven:mutationCoverage
+{% endhighlight %}
+
+It will output an html report to *target/pit-reports/YYYYMMDDHHMI*. In this report we can see that we have 100% line coverage, but only 75% mutation coverage. It means that our tests are not as good as they should be.
+
+<img src="/images/blog/posts/pitest-with-java/lineCoverage.jpg" />
+
+When you'll click on class name you'll see a detail report.
```
Comment (natif) [inline _posts/2016-05-30-pitest-with-java.md]: when you click on class name you will see a detail report

File: _posts/2016-05-30-pitest-with-java.md [hunk 10]
```diff
@@ -0,0 +1,263 @@
...
+
+<img src="/images/blog/posts/pitest-with-java/detailReport.jpg" />
+
+We have one type of mutation which survived, it's conditional boundary change. We already know that > operator is mutated to >=. Mutation survived, because when we'll change > to >= our tests will pass.
```
Comment (natif) [inline _posts/2016-05-30-pitest-with-java.md]: We already know, (coma)

File: _posts/2016-05-30-pitest-with-java.md [hunk 11]
```diff
@@ -0,0 +1,263 @@
...
+	Passenger adultPassenger1 = new Passenger(20);
+	Passenger childPassenger2 = new Passenger(12);
+	passengers.add(adultPassenger1);
+	passengers.add(childPassenger2);
+	double price = calculator.calculatePrice(passengers, ADULT_TICKET_PRICE, CHILD_TICKER_PRICE);
+	assertEquals((ADULT_TICKET_PRICE + CHILD_TICKER_PRICE), price, 0);
+}
+{% endhighlight %}
+
+Ok, we covered our edge cases, however if you'll execute pitest plugin once again you'll see that we still have two mutations which survived.
```
Comment (natif) [inline _posts/2016-05-30-pitest-with-java.md]: if you execute (without will)

File: _posts/2016-05-30-pitest-with-java.md [hunk 12]
```diff
@@ -0,0 +1,263 @@
...
+Our edge case test scenario passes one adult and one child to this method. Due to AND operator we can change our code to:
+{% highlight java %}
+if (childrenCounter >= 1 && adultCounter > 1) {
+    result = (1 - FAMILY_DISCOUNT) * totalPrice;
+} else {
+    result = totalPrice;
+}
+{% endhighlight %}
+
+and all test all still green, because second condition *adultCounter > 1* change the result of this statement to false. The same behaviour we can observe when we'll change *adultCounter > 1* to *adultCounter >= 1*.
```
Comment (natif) [inline _posts/2016-05-30-pitest-with-java.md]: and all test are still green
1\* changed
we can -> might be observed

File: _posts/2016-05-30-pitest-with-java.md [hunk 13]
```diff
@@ -0,0 +1,263 @@
...
+	passengers.add(childPassenger2);
+	double price = calculator.calculatePrice(passengers, ADULT_TICKET_PRICE, CHILD_TICKER_PRICE);
+	assertEquals((ADULT_TICKET_PRICE + 2 * CHILD_TICKER_PRICE), price, 0);
+}
+{% endhighlight %}
+
+After this we finally reached 100% mutation coverage.
+
+# Summary
+Now you know how to use mutation testing to see if your tests are good. You also know that line coverage doesn't mean that we have covered all cases. Hope you enjoyed and you'll start to use mutation testing. All code which was used is available on [repository](https://github.com/BeyondScheme/pitest-java-example).
```
Comment (natif) [inline _posts/2016-05-30-pitest-with-java.md]: Now you know - should be changed. Maybe -> To sum up, As shown above...

Comment (natif) [inline _posts/2016-05-30-pitest-with-java.md]: The code which was used is available at Beyond Scheme's repository (link the repository).

Comment (Tomabana) [inline _posts/2016-05-30-pitest-with-java.md]: To sum up, after this reading you should learn how to use mutation testing to see if your tests are good.

**Comment by ArturT** [comment]
You can add README.md in the repository https://github.com/BeyondScheme/pitest-java-example and the link to blog post in it. Also please add the footer with our brand, similar as here https://github.com/BeyondScheme/elixir-game_of_life#about-beyond-scheme

File: _posts/2016-05-30-pitest-with-java.md (hunk 2 above)
Comment (ArturT) [inline _posts/2016-05-30-pitest-with-java.md]: "... how to use mutation testing in Java to test unit tests."
I would mention we are talking about mutation testing in Java. More keywords for google and it's clear for readers.

File: _posts/2016-05-30-pitest-with-java.md [hunk 14]
```diff
@@ -0,0 +1,263 @@
+---
+layout: post
+title:  "Test your tests with pitest"
```
Comment (ArturT) [inline _posts/2016-05-30-pitest-with-java.md]: I would add word Java somewhere in the title.

File: _posts/2016-05-30-pitest-with-java.md [hunk 15]
```diff
@@ -0,0 +1,263 @@
...
+
+|Return type |Mutation|
+|:-------------:|:-----:|
+|boolean|change true to false, false to true|
+|int byte short|change 0 to 1, otherwise return 0|
+|long|change x to x+1|
+
+There are few libraries which can be used to mutation testing. I chose [pitest](http://pitest.org/) library.
+
+# Usage
```
Comment (ArturT) [inline _posts/2016-05-30-pitest-with-java.md]: This title is super short, would be nice to make it more descriptive. Usage of what? Maybe name it differently? That's my first thought when I saw it.

File: _posts/2016-05-30-pitest-with-java.md (hunk 5 above)
Comment (ArturT) [inline _posts/2016-05-30-pitest-with-java.md]: I like when names of classes or variables, all stuff related to the code are presented in `Passenger` instead of _italic_. It seems to more readable for me. 

```
`Presenter` instead of *Presenter*
```

File: _posts/2016-05-30-pitest-with-java.md (hunk 14 above)
Comment (Tomabana) [inline _posts/2016-05-30-pitest-with-java.md]: I changed to "Pitest - mutation testing in Java"

File: _posts/2016-05-30-pitest-with-java.md (hunk 15 above)
Comment (Tomabana) [inline _posts/2016-05-30-pitest-with-java.md]: I changed title to "Example of usage pitest in Java"
